### ISSUE TITLE
Evolving Zelda Graph Grammar for Dungeons

### ISSUE THREAD

**Issue by jakegut** [schrum2/MM-NEAT#342]
There are multiple ways to implement this:
One is to use FI-2 Pop and generate dungeons based on genetic algorithm or generate a "grammar" for a dungeon and then generate the levels based on the "grammar" (#341)

**Comment by schrum2** [comment]
Here is a highly relevant paper you should read:
http://antoniosliapis.com/papers/multi-segment_evolution_of_dungeon_game_levels.pdf

**Comment by schrum2** [comment]
I changed the name of this issue. We can already generate dungeons using a grammar, but we will make this particular issue correspond to evolving the dungeons (perhaps also with a grammar). This is low priority for now though ... not sure we'll get to it.

**Comment by schrum2** [comment]
The CPPN2GAN paper evolved dungeons by first evolving CPPNs and using them to generate dungeons with a GAN. However, I'm keeping this issue around to specifically focus on evolving the graph grammar used to generate the dungeons. So, this issue primarily focuses on the encoding.

**Comment by schrum2** [comment]
This is a long-term issue. There are many others to consider first.

**Comment by schrum2** [comment]
Here is some stuff to start looking at:
https://cran.r-project.org/web/packages/gramEvol/vignettes/ge-intro.pdfhttps://www.researchgate.net/publication/262216512_GEVA_grammatical_evolution_in_Java
http://ncra.ucd.ie/Site/GEVA.html

These are meant to give you an idea of what grammatical evolution is. Hopefully, once you have some understanding there, we can move on to taking our own existing grammar rules and evolving dungeons.

Basically, the idea is that we specify the grammar rules, but instead of randomly selecting options from the rule set, evolution figures out what suits our design needs (fitness function) the most. I suppose we'll also need to evolve the dungeon backbones as well. Ultimately, this will involve extending ZeldaDungeonTask, and for that reason along it might be a good idea to do #457 first

**Comment by schrum2** [comment]
Not really interested in evolving the grammar. Also, the code for the graph grammar is a bit of a tricky mess any way.
